src/backend/distributed/executor/multi_server_executor.c

```diff
@@ -74,37 +75,69 @@ JobExecutorType(DistributedPlan *distributedPlan)
 		int dependedJobCount = 0;
 
 		/* if we need to open too many connections per worker, warn the user */
-		if (tasksPerNode >= MaxConnections)
+		if (tasksPerNode >= MaxConnections && !AllowExecutorSwap)
 		{
 			ereport(WARNING, (errmsg("this query uses more connections than the "
 									 "configured max_connections limit"),
-							  errhint("Consider increasing max_connections or setting "
+							  errhint("Consider increasing max_connections, setting "
 									  "citus.task_executor_type to "
-									  "\"task-tracker\".")));
+									  "\"task-tracker\" or "
+									  "citus.allow_executor_swap to ON.")));
 		}
+		else if (taskCount >= reasonableConnectionCount && AllowExecutorSwap)
+		{
+			ereport(WARNING, (errmsg("this query uses more connections than the "
+									 "configured max_connections limit"),
+							  errhint("Since you enabled citus.allow_executor_swap "
+									  "Citus chose to use task-tracker. "
+									  "Consider increasing max_connections to run "
+									  "this query faster.")));
+			return MULTI_EXECUTOR_TASK_TRACKER;
+		}
+
 
 		/*
 		 * If we need to open too many outgoing connections, warn the user.
 		 * The real-time executor caps the number of tasks it starts by the same limit,
 		 * but we still issue this warning because it degrades performance.
 		 */
 		reasonableConnectionCount = MaxMasterConnectionCount();
-		if (taskCount >= reasonableConnectionCount)
+		if (taskCount >= reasonableConnectionCount && !AllowExecutorSwap)
 		{
 			ereport(WARNING, (errmsg("this query uses more file descriptors than the "
 									 "configured max_files_per_process limit"),
 							  errhint("Consider increasing max_files_per_process or "
 									  "setting citus.task_executor_type to "
-									  "\"task-tracker\".")));
+									  "\"task-tracker\" or "
+									  "citus.allow_executor_swap to ON.")));
+		}
+		else if (taskCount >= reasonableConnectionCount && AllowExecutorSwap)
+		{
+			ereport(WARNING, (errmsg("this query uses more file descriptors than the "
+									 "configured max_files_per_process limit"),
+							  errhint("Since you enabled citus.allow_executor_swap "
+									  "Citus chose to use task-tracker. "
+									  "Consider increasing max_files_per_process to run "
+									  "this query faster.")));
+			return MULTI_EXECUTOR_TASK_TRACKER;
```

Comment:
We don't quite want this right? This *increases* the cases where we use the task-tracker executor. Before this we would warn, but still use the real-time executor. We should only chose the task-tracker executor for repartition jobs dynamically. We then probably want a separate config to turn off task-tracker entirely, which would retain current behavior.

Reply:
Right, we should only consider switching when dependedJobCount > 0.